torchtune/datasets/_chat.py

```diff
@@ -0,0 +1,181 @@
+
+    Use `convert_to_dialogue` to prepare your dataset into the llama conversation format
+    and roles:
+        [
+            {
+                "role": <system|user|assistant>,
+                "content": <message>,
+            },
+            ...
+        ]
+
+    This class supports multi-turn conversations. If a tokenizer sample with multiple
+    turns does not fit within `max_seq_len` then it is truncated.
+
+    Args:
+        tokenizer (Tokenizer): Tokenizer used to encode data. Tokenize must implement an `encode` and `decode` method.
+        source (str): path string of dataset, anything supported by HuggingFace's `load_dataset`
+            (https://huggingface.co/docs/datasets/en/package_reference/loading_methods#datasets.load_dataset.path)
+        convert_to_dialogue (Callable[[Mapping[str, Any]], Dialogue]): function that keys into the desired field in the sample
+            and converts to a list of `Messages` that follows the llama format with the expected keys
+        template (PromptTemplate): template used to format the prompt. If the placeholder variable
+            names in the template do not match the column/key names in the dataset, use `column_map` to map them.
+        max_seq_len (int): Maximum number of tokens in the returned input and label token id lists.
+        train_on_input (bool): Whether the model is trained on the prompt or not. Default is False.
+        **load_dataset_kwargs (Dict[str, Any]): additional keyword arguments to pass to `load_dataset`.
+    """
+
+    def __init__(
+        self,
+        tokenizer: Tokenizer,
+        source: str,
+        convert_to_dialogue: Callable[[Mapping[str, Any]], Dialogue],
+        template: PromptTemplate,
+        max_seq_len: int,
+        train_on_input: bool = False,
+        **load_dataset_kwargs: Dict[str, Any],
+    ) -> None:
+        self._tokenizer = tokenizer
+        self._data = load_dataset(source, **load_dataset_kwargs)
+        self._convert_to_dialogue = convert_to_dialogue
+        self.template = template
+        self.max_seq_len = max_seq_len
+        self.train_on_input = train_on_input
+
+    def __len__(self):
+        return len(self._data)
+
+    def __getitem__(self, index: int) -> Tuple[List[int], List[int]]:
+        sample = self._data[index]
+        return self._prepare_sample(sample)
+
+    def _prepare_sample(self, sample: Mapping[str, Any]) -> Tuple[List[int], List[int]]:
+        dialogue = self._convert_to_dialogue(sample)
+
+        prompt_tokens = []
+        label_tokens = []
+        for prompt, label in self._get_turns(dialogue):
+            formatted_prompt = self.template.format(prompt)
+            encoded_prompt_with_response, labels = tokenize_prompt_and_response(
+                tokenizer=self._tokenizer,
+                prompt=formatted_prompt,
+                response=label,
+                train_on_input=self.train_on_input,
+            )
```

Comment:
We need to think about whether we can make this more general

Reply:
yeah - let's tokenize an individual prompt or response. let me rethink this